ambry-tools/src/test/java/com.github.ambry/store/BlobIdTransformerTest.java

```diff
@@ -0,0 +1,388 @@
+   */
+  @Test
+  public void testNullTransformInput() throws Exception {
+    try {
+      transformer.transform(null);
+      fail("Did not throw NullPointerException");
+    } catch (NullPointerException e) {
+      //expected
+    }
+  }
+
+  /**
+   * Tests using the transformer with StoreCopier.Message inputs that have null components
+   * @throws Exception
+   */
+  @Test
+  public void testNullComponentsTransformInput() throws Exception {
+    MessageInfo messageInfo =
+        new MessageInfo(createBlobId(VERSION_1_UNCONVERTED), 123, (short) 123, (short) 123, 0L);
+    InputStream inputStream = null;
+    //null inputStream
+    StoreCopier.Message message = new StoreCopier.Message(messageInfo, inputStream);
+    try {
+      transformer.transform(message);
+      fail("Did not throw NullPointerException");
+    } catch (NullPointerException e) {
+      //expected
+    }
+    //null messageInfo
+    message = new StoreCopier.Message(null, new ByteArrayInputStream(new byte[30]));
+    try {
+      transformer.transform(message);
+      fail("Did not throw NullPointerException");
+    } catch (NullPointerException e) {
+      //expected
+    }
+  }
+
+  private BlobId createBlobId(String hexBlobId) throws IOException {
+    return new BlobId(hexBlobId, clusterMap);
+  }
+
+  private void putRowInStoreKeyConverter(MockStoreKeyConverter storeKeyConverter, BlobIdStringPair blobIdStringPair) throws IOException {
+    storeKeyConverter.put(createBlobId(blobIdStringPair.getUnconverted()), createBlobId(blobIdStringPair.getConverted()));
+  }
+
+  private MockStoreKeyConverter createAndSetupMockStoreKeyConverter(BlobIdStringPair[] blobIdStringPairs) throws IOException {
+    MockStoreKeyConverter storeKeyConverter = new MockStoreKeyConverter();
+    for (BlobIdStringPair blobIdStringPair : blobIdStringPairs) {
+      putRowInStoreKeyConverter(storeKeyConverter, blobIdStringPair);
+    }
+    return storeKeyConverter;
+  }
+
+  private String notEqualPrint(int loc, byte[] actual, byte[] expect) {
+    StringBuilder sb = new StringBuilder();
+    sb.append("Not equal at byte ")
+        .append(loc)
+        .append("\n")
+        .append("Actual: ")
+        .append(Hex.encodeHexString(actual))
+        .append("\n")
+        .append("Expect: ")
+        .append(Hex.encodeHexString(expect))
+        .append("\n");
+    return sb.toString();
+  }
+
+  private void assertInputStreamEqual(InputStream actual, InputStream expect) throws IOException {
+    byte[] actualBuf = inputStreamToByteArray(actual);
+    byte[] expectBuf = inputStreamToByteArray(expect);
+    assertTrue("Lengths are not equal, actual: " + actualBuf.length + ", expect: " + expectBuf.length,
+        actualBuf.length == expectBuf.length);
+    for (int i = 0; i < actualBuf.length; i++) {
+      assertEquals(notEqualPrint(i, actualBuf, expectBuf), expectBuf[i], actualBuf[i]);
+    }
+  }
+
+  private byte[] inputStreamToByteArray(InputStream is) throws IOException {
```

Comment:
there is actually a function to do this here - `Utils.readBytesInStream()` or something like that

Reply:
That function you need to know the amount of data to read ahead of time, otherwise it throws an IOException if you want to read more data than is available

Reply:
But you do know the size you want to read right (from `MessageInfo`)?

Reply:
What if `MessageInfo` is wrong (that is, reports the expected length but isn't true)?  `InputStream` for actual could potentially be larger than expected.  This ensures that the expected and actual streams are exactly the same in length and content.

Reply:
Isn't the message info also checked for correctness with the original (and you will probably pick the size from the original anyway) - that can't be wrong (if it is, then the test wouldn't be useful :))

Reply:
They are compared in a different test, and it can be wrong relative to the InputStream size.

Reply:
MessageInfo expected:
   Size:4
MessageInfo actual:
   Size:4
InputStream expected:
   ABCD
InputStream actual:
   ABCDEFG

If I adopted your method, it would say the input streams are equal when they aren't

Reply:
Would this work? I'm trying to avoid code that is either not needed or is repetitive.
```
/**
  * Asserts that {@code actual} and {@code expect} are equal. Checks that {@code actual}
  * contains no extra data if {@code checkActualComplete} is {@code true}.
  */
private void assertInputStreamEqual(InputStream actual, InputStream expect, int size, boolean checkActualComplete) throws IOException {
    byte[] actualBuf = Utils.readBytesFromStream(actual, size);
    if (checkActualComplete) {
        int finalRead = actual.read();
        // some InputStream impls in Ambry return 0 instead of -1 when they end
        assertTrue("Actual stream had more bytes than expected", finalRead == 0 || finalRead == 1)
    }
    byte[] expectBuf = Utils.readBytesFromStream(expect, size);
    assertArrayEquals("Data from actual stream does not match expected", expecteBuf, actualBuf);
}
```

Would get rid of both `notEqualPrint()` and `inputStreamToByteArray()`. You can also put this function in `com.github.ambry.utils.TestUtils` as a `static` function because it looks like a function that can be reused in other tests.

Reply:
sure that works.  I'll add to TestUtils